tests/semantic_analyser.cpp

```diff
@@ -3103,19 +3105,43 @@ TEST(semantic_analyser, tuple)
   test(R"_(BEGIN { @t = (1, 2, "string"); @t = (3, 4, "other"); })_");
   test(R"_(BEGIN { @t = (1, kstack()) })_");
   test(R"_(BEGIN { @t = (1, (2,3)) })_");
+  test(R"_(BEGIN { $t = (1, (int64)2); $t = (2, (int32)3); })_");
+
+  test_error(R"_(BEGIN { $t = (1, (int32)2); $t = (2, (int64)3); })_", R"(
+stdin:1:29-47: ERROR: Tuple type mismatch: (int64,int32) != (int64,int64).
+BEGIN { $t = (1, (int32)2); $t = (2, (int64)3); }
+                            ~~~~~~~~~~~~~~~~~~
+)");
 
   test(R"_(struct task_struct { int x; } BEGIN { $t = (1, curtask); })_");
   test(R"_(struct task_struct { int x[4]; } BEGIN { $t = (1, curtask->x); })_");
 
   test(R"_(BEGIN { $t = (1, 2); $t = (4, "other"); })_", 10);
   test(R"_(BEGIN { $t = (1, 2); $t = 5; })_", 1);
   test(R"_(BEGIN { $t = (1, count()) })_", 1);
-  test(R"_(BEGIN { $t = ((int32)1, (int64)2); $t = ((int64)1, (int32)2); })_",
-       10);
 
   test(R"_(BEGIN { @t = (1, 2); @t = (4, "other"); })_", 10);
   test(R"_(BEGIN { @t = (1, 2); @t = 5; })_", 1);
   test(R"_(BEGIN { @t = (1, count()) })_", 1);
+
+  test(R"_(BEGIN { $t = (1, (2, 3)); $t = (4, ((int8)5, 6)); })_");
+
+  test_error(R"_(BEGIN { $t = (1, ((int8)2, 3)); $t = (4, (5, (int8)6)); })_",
```

Comment:
I'm a bit lost here. Why does the first one pass but the second one trigger an error?

Reply:
Because the '5' in the first position (of the nested tuple) of the second assignment doesn't fit into the `int8` in the same position in the first assignment. The second int8 cast for the 6 is kinda superfluous, I'll remove it.

Reply:
Right, that's probably what confused me. Thanks for the explanation!